File: ios/LocalPods/ScannerModule/Core/Network/EvidenceAPI.swift
// LocalPods/ScannerModule/Core/Network/EvidenceAPI.swift
//
// Evidence API Client
// Handles /evidences endpoints (§6, §7, §8)

import Foundation
import UIKit

/// Evidence API Client for multipart uploads
final class EvidenceAPI: BaseAPIClient {

    /// POST /evidences/ingest (§6.2)
    /// Multipart upload with image + metadata
    func ingest(
        treeId: String,
        imageId: String,
        image: UIImage,
        timeSeries: TimeSeriesData,
        metadata: MetadataData
    ) async throws -> IngestResponse {
        // Convert image to JPEG data.
        // Power-aware (field test 2026-05-15): 0.8 → 0.6 — file giảm ~40%,
        // upload time giảm tương ứng, radio bật ngắn hơn → mát máy hơn.
        // Visual quality vẫn đủ cho AI training (yolov26seg detect chính xác trên 0.6).
        guard let imageData = image.jpegData(compressionQuality: 0.6) else {
            throw ApiError.invalidResponse
        }

        // Build payload JSON
        let payload: [String: Any] = [
            "image_id": imageId,
            "tree_id": treeId,
            "time_series": [
                "latitude": timeSeries.latitude,
                "longitude": timeSeries.longitude,
                "timestamp": timeSeries.timestamp,
                "heading": timeSeries.heading,
                "pitch": timeSeries.pitch,
                "roll": timeSeries.roll
            ],
            "metadata": [
                "device_id": metadata.deviceId,
                "nonce": metadata.nonce,
                "signature": metadata.signature
            ]
        ]

        let payloadData = try JSONSerialization.data(withJSONObject: payload)
        guard let payloadString = String(data: payloadData, encoding: .utf8) else {
            throw ApiError.invalidResponse
        }

        // Create multipart request
        let boundary = "Boundary-\(UUID().uuidString)"
        var request = createRequest(path: "/evidences/ingest", method: "POST")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // Add payload field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"payload\"\r\n\r\n".data(using: .utf8)!)
        body.append(payloadString.data(using: .utf8)!)
        body.append("\r\n".data(using: .utf8)!)

        // Add image field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"image\"; filename=\"image.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)

        // End boundary
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        // Execute request
        return try await execute(request: request, responseType: IngestResponse.self)
    }
}

// MARK: - Models

/// Time series data (§6.1 TimeSeriesForm)
struct TimeSeriesData {
    let latitude: Double
    let longitude: Double
    let timestamp: Int64  // Unix epoch MILLISECONDS
    let heading: Double
    let pitch: Double
    let roll: Double
}

/// Metadata for device signature (§6.1 MetadataForm)
struct MetadataData {
    let deviceId: String
    let nonce: String
    let signature: String
}

/// Ingest response (§6.2)
struct IngestResponse: Codable {
    let success: Bool
    let imageId: String
    let treeId: String
    let featuresExtracted: FeaturesExtracted?
    let storageKeys: StorageKeys?
    let message: String

    enum CodingKeys: String, CodingKey {
        case success
        case imageId = "image_id"
        case treeId = "tree_id"
        case featuresExtracted = "features_extracted"
        case storageKeys = "storage_keys"
        case message
    }
}

struct FeaturesExtracted: Codable {
    let globalDim: Int
    let localKeypoints: Int
    let localDim: Int

    enum CodingKeys: String, CodingKey {
        case globalDim = "global_dim"
        case localKeypoints = "local_keypoints"
        case localDim = "local_dim"
    }
}

struct StorageKeys: Codable {
    let globalFeatures: String?
    let localFeatures: String?
    let image: String?

    enum CodingKeys: String, CodingKey {
        case globalFeatures = "global_features"
        case localFeatures = "local_features"
        case image
    }
}

File: android/orilifesdk/src/main/java/com/mvp/orilife/network/EvidenceAPI.kt
package com.mvp.orilife.network

import android.util.Log
import com.google.gson.Gson
import com.mvp.orilife.network.ApiEnvelope
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Evidence API Client
 * Handles /evidences/ingest endpoint (§6.2)
 */
class EvidenceAPI(
    private val baseUrl: String,
    private val apiKey: String
) {
    private val client: OkHttpClient
    private val gson = Gson()

    init {
        val loggingInterceptor = HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        }

        client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(loggingInterceptor)
            .build()

        Log.d(TAG, "✅ EvidenceAPI initialized: baseUrl=$baseUrl")
    }

    /**
     * POST /evidences/ingest - Ingest evidence (§6.2)
     * Multipart upload with image + metadata
     */
    suspend fun ingest(request: IngestRequest): Result<IngestResponse> {
        return try {
            Log.d(TAG, "📤 POST /evidences/ingest")
            Log.d(TAG, "   → treeId: ${request.treeId}")
            Log.d(TAG, "   → imageId: ${request.imageId}")
            Log.d(TAG, "   → image: ${request.imageFile.name} (${request.imageFile.length()} bytes)")

            // Detect MIME type
            val mimeType = when {
                request.imageFile.name.lowercase().endsWith(".png") -> "image/png"
                else -> "image/jpeg"
            }
            val imageBody = request.imageFile.asRequestBody(mimeType.toMediaType())

            // Build payload JSON (iOS format: 2 fields instead of 5)
            val payloadJson = buildString {
                append("{")
                append("\"image_id\":\"${request.imageId}\"")
                append(",\"tree_id\":\"${request.treeId}\"")
                append(",\"time_series\":{")
                append("\"latitude\":${request.timeSeries.latitude ?: "null"}")
                append(",\"longitude\":${request.timeSeries.longitude ?: "null"}")
                append(",\"timestamp\":${request.timeSeries.timestamp}")
                append(",\"heading\":${request.timeSeries.heading ?: "null"}")
                append(",\"pitch\":${request.timeSeries.pitch ?: "null"}")
                append(",\"roll\":${request.timeSeries.roll ?: "null"}")
                append("}")
                append(",\"metadata\":{")
                append("\"device_id\":\"${request.metadata.deviceId}\"")
                append(",\"nonce\":\"${request.metadata.nonce}\"")
                append(",\"signature\":\"${request.metadata.signature}\"")
                append("}")
                append("}")
            }

            val requestBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("payload", payloadJson)
                .addFormDataPart("image", request.imageFile.name, imageBody)
                .build()

            val httpRequest = Request.Builder()
                .url("$baseUrl/evidences/ingest")
                .header("X-API-Key", apiKey)
                .post(requestBody)
                .build()

            val response = client.newCall(httpRequest).execute()
            val responseBody = response.body?.string() ?: ""

            if (!response.isSuccessful) {
                Log.e(TAG, "❌ POST /evidences/ingest failed: ${response.code} - $responseBody")
                return Result.failure(ApiException(response.code, responseBody))
            }

            val envelope = gson.fromJson(responseBody, ApiEnvelope::class.java)
            val ingestResponse = gson.fromJson(gson.toJson(envelope.data), IngestResponse::class.java)

            Log.d(TAG, "✅ POST /evidences/ingest success: imageId=${ingestResponse.imageId}, treeId=${ingestResponse.treeId}")
            Result.success(ingestResponse)

        } catch (e: Exception) {
            Log.e(TAG, "❌ POST /evidences/ingest error: ${e.message}", e)
            Result.failure(e)
        }
    }

    companion object {
        private const val TAG = "EvidenceAPI"
    }
}

// MARK: - Request Models

data class IngestRequest(
    val treeId: String,
    val imageId: String,
    val imageFile: File,
    val timeSeries: TimeSeriesData,
    val metadata: MetadataData
)

data class TimeSeriesData(
    val latitude: Double?,
    val longitude: Double?,
    val timestamp: Long,
    val heading: Double?,
    val pitch: Double?,
    val roll: Double?
)

data class MetadataData(
    val deviceId: String,
    val nonce: String,
    val signature: String
)

// MARK: - Response Models

data class IngestResponse(
    val success: Boolean,
    val imageId: String,
    val treeId: String,
    val featuresExtracted: FeaturesExtracted?,
    val storageKeys: StorageKeys?,
    val message: String
)

data class FeaturesExtracted(
    val globalDim: Int,
    val localKeypoints: Int,
    val localDim: Int
)

data class StorageKeys(
    val globalFeatures: String?,
    val localFeatures: String?,
    val image: String?
)
